File: packages/swift/Sources/VidcallWebRTC/PeerConnectionManager.swift
//
//  PeerConnectionManager.swift
//  VidcallWebRTC
//
//  Peer connection + offer/answer/ICE wired to the vidcall signaling protocol
//  (protocol/schema.json). SDP (`OfferPayload`) and ICE (`IcePayload`) are
//  relayed verbatim — the signaling layer never parses or transforms them.
//
//  The negotiation state machine implements the "perfect negotiation" pattern
//  (same as packages/core/src/peer-connection-manager.ts):
//
//  - **Glare**: on a colliding remote offer, the polite peer rolls back its
//    own *local* offer (`setLocalDescription(.rollback)`) and accepts the
//    remote one; the impolite peer ignores the colliding remote offer.
//  - **Trickle ICE (RFC 8838)**: local candidates are signaled as they fire;
//    remote candidates are queued until the matching remote description has
//    been applied, then flushed (spec requires `addIceCandidate` after
//    `setRemoteDescription`).
//  - **ICE restart (RFC 8445 §9)**: `restartIce()` triggers a fresh offer
//    with new ICE credentials; `autoRestartIce` does the same automatically
//    when `iceConnectionState` turns `failed`.
//  - **Data channels**: the bus creates the local channel (included in the
//    next offer); a remote channel delivered via `onDataChannel` is adopted
//    and becomes active (mirrors the TS `DataChannelBus`).
//
//  The manager is WebRTC-agnostic — it drives a `PeerConnectionSession`
//  (real adapter: `RTCPeerConnectionSession`; tests inject fakes).
//

import Foundation
import Vidcall

/// Outbound signaling hook: the manager emits offer/answer/ice wire payloads.
/// The real implementation wraps `VidcallClient`; tests use an in-process
/// signaling bridge.
public protocol PeerSignaling: AnyObject, Sendable {
    func sendOffer(_ payload: OfferPayload) throws
    func sendAnswer(_ payload: OfferPayload) throws
    func sendIce(_ payload: IcePayload) throws
}

/// Connection status of the underlying peer connection.
public enum PeerConnectionStatus: Equatable, Sendable {
    case idle
    case connecting
    case connected
    case disconnected
    case failed(String)
}

/// WebRTC-agnostic surface of the peer-connection manager. `VidcallClient`
/// events are fed through `handleIncoming(event:)`; SDP/ICE flow back out via
/// the `PeerSignaling` hook.
public protocol PeerConnectionManaging: AnyObject, Sendable {
    /// Feeds a decoded client event into the negotiation state machine.
    func handleIncoming(event: VidcallClient.Event)
    /// Applies a remote trickle ICE candidate (type `ice`).
    func addRemoteIceCandidate(_ payload: IcePayload) throws
    /// Ensures an offer/answer exchange (perfect negotiation).
    func negotiate() async throws
    /// Restarts ICE with fresh credentials (RFC 8445 §9).
    func restartIce() async throws
    /// The per-connection typed data-channel bus (created eagerly so the
    /// local data channel is part of the first offer).
    var dataChannelBus: DataChannelBus? { get }
    /// Closes the peer connection and removes this manager from the client.
    func leave()
}

/// Concrete peer-connection manager: the negotiation state machine over a
/// `PeerConnectionSession`. Create the real WebRTC-backed instance via
/// `VidcallWebRTC.makePeerConnectionManager`; tests construct it directly
/// with a fake session and an in-process signaling bridge.
public final class PeerConnectionManager: PeerConnectionManaging, @unchecked Sendable {
    /// Stream id used for local media tracks.
    public static let localStreamId = "vidcall-stream"
    /// Default data-channel label (matches the TS core `'vidcall'`).
    public static let defaultDataChannelName = "vidcall"

    // MARK: Callbacks

    /// A local SDP offer is ready — forward via `sendOffer` (already emitted
    /// through `signaling`; use for observability/relay).
    public var onLocalOffer: (@Sendable (OfferPayload) -> Void)?
    /// A local SDP answer is ready.
    public var onLocalAnswer: (@Sendable (OfferPayload) -> Void)?
    /// A local trickle ICE candidate is ready.
    public var onLocalIceCandidate: (@Sendable (IcePayload) -> Void)?
    /// Status changes of the peer connection.
    public var onStatusChange: (@Sendable (PeerConnectionStatus) -> Void)?

    // MARK: Public state

    /// The session (real or fake) this manager drives.
    public let session: PeerConnectionSession
    /// Effective configuration.
    public let configuration: PeerConnectionConfiguration
    /// Outbound signaling hook.
    public let signaling: PeerSignaling
    /// Optional remote peer id — when set, only envelopes from this sender are
    /// accepted (single-peer safety against broadcast backends).
    public let remotePeerId: String?

    /// Current status.
    public private(set) var status: PeerConnectionStatus = .idle {
        didSet {
            guard oldValue != status else { return }
            onStatusChange?(status)
        }
    }

    // MARK: Private state

    private let lock = NSLock()

    /// Runs `body` while holding the state lock. All mutable manager state is
    /// accessed through this helper so the async negotiation paths never call
    /// `NSLock` methods directly (Swift 6 concurrency safety).
    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    private var makingOffer = false
    private var ignoreRemoteOffer = false
    private var restartingIce = false
    private var closed = false
    private var pendingIceCandidates: [IceCandidate] = []
    private var _dataChannelBus: DataChannelBus?

    // MARK: Init

    /// Wires the manager to `session` and registers the session callbacks.
    public init(
        session: PeerConnectionSession,
        configuration: PeerConnectionConfiguration,
        signaling: PeerSignaling,
        remotePeerId: String? = nil
    ) {
        self.session = session
        self.configuration = configuration
        self.signaling = signaling
        self.remotePeerId = remotePeerId
        session.onLocalIceCandidate = { [weak self] candidate in
            self?.handleLocalIceCandidate(candidate)
        }
        session.onDataChannel = { [weak self] channel in
            self?.handleRemoteDataChannel(channel)
        }
        session.onNegotiationNeeded = { [weak self] in
            self?.negotiationNeeded()
        }
        session.onConnectionStateChange = { [weak self] state in
            self?.handleConnectionState(state)
        }
        // The bus is created eagerly so the local data channel exists before
        // the first offer (it is part of the negotiated SCTP setup — same as
        // the TS core, whose `Room` creates the bus at construction).
        _dataChannelBus = DataChannelBus(session: session, name: PeerConnectionManager.defaultDataChannelName)
    }

    deinit {
        session.close()
    }

    // MARK: PeerConnectionManaging

    public func handleIncoming(event: VidcallClient.Event) {
        switch event {
        case .offer(let envelope, let payload):
            guard accepts(senderId: envelope.senderId) else { return }
            Task { [weak self] in
                do {
                    try await self?.receiveRemoteOffer(payload)
                } catch {
                    self?.status = .failed(error.localizedDescription)
                }
            }
        case .answer(let envelope, let payload):
            guard accepts(senderId: envelope.senderId) else { return }
            Task { [weak self] in
                do {
                    try await self?.receiveRemoteAnswer(payload)
                } catch {
                    self?.status = .failed(error.localizedDescription)
                }
            }
        case .ice(let envelope, let payload):
            guard accepts(senderId: envelope.senderId) else { return }
            do {
                try addRemoteIceCandidate(payload)
            } catch {
                status = .failed(error.localizedDescription)
            }
        default:
            break
        }
    }

    public func addRemoteIceCandidate(_ payload: IcePayload) throws {
        guard !withLock({ closed }) else { return }
        let candidate = IceCandidate(
            sdp: payload.candidate,
            sdpMid: payload.sdpMid,
            sdpMLineIndex: Int32(payload.sdpMLineIndex ?? 0)
        )
        if session.remoteDescriptionIsSet {
            addCandidate(candidate)
        } else {
            // Trickle: queue until the remote description lands (RFC 8838).
            withLock {
                pendingIceCandidates.append(candidate)
            }
        }
    }

    public func negotiate() async throws {
        try await negotiate(iceRestart: false)
    }

    /// Restarts ICE: a new offer with fresh ufrag/pwd is negotiated (RFC 8445
    /// §9), i.e. `negotiate()` with the stack's ICE-restart path enabled
    /// (`RTCPeerConnection.restartIce()` + `IceRestart` constraints in the
    /// real adapter; fakes implement `createOffer(iceRestart:)` directly).
    public func restartIce() async throws {
        let proceed = withLock { () -> Bool in
            guard !closed, !restartingIce else { return false }
            restartingIce = true
            return true
        }
        guard proceed else { return }
        defer {
            withLock { restartingIce = false }
        }
        try await negotiate(iceRestart: true)
    }

    /// Core negotiation. Guards against overlapping offers; the local data
    /// channel is created before the offer so it is part of the SCTP setup.
    private func negotiate(iceRestart: Bool) async throws {
        let proceed = withLock { () -> Bool in
            guard !closed, !makingOffer else { return false }
            makingOffer = true
            return true
        }
        guard proceed else { return }
        defer {
            withLock { makingOffer = false }
        }

        // The local data channel must exist before the offer so it is part of
        // the negotiated SCTP setup (mirrors the TS bus).
        ensureDataChannelBus()

        let offer = try await session.createOffer(iceRestart: iceRestart)
        try await session.setLocalDescription(offer)
        let payload = OfferPayload(sdp: offer.sdp, label: iceRestart ? "ice-restart" : "main")
        onLocalOffer?(payload)
        try signaling.sendOffer(payload)
    }

    /// The typed data-channel bus for this peer (created eagerly in `init`).
    public var dataChannelBus: DataChannelBus? {
        withLock { _dataChannelBus }
    }

    public func leave() {
        withLock {
            guard !closed else { return }
            closed = true
        }
        session.close()
        withLock { pendingIceCandidates.removeAll() }
        status = .disconnected
    }

    // MARK: Remote signaling

    // Internal (not private) so in-process signaling bridges / tests can
    // drive the state machine directly and await completion.

    func receiveRemoteOffer(_ payload: OfferPayload) async throws {
        guard !withLock({ closed }) else { return }
        let sdp = SessionDescription(type: .offer, sdp: payload.sdp)

        let collision = withLock {
            makingOffer || session.signalingState != .stable
        }

        if collision {
            if !configuration.polite {
                // Impolite side: ignore the colliding remote offer (perfect
                // negotiation). The remote side will roll back and answer ours.
                withLock { ignoreRemoteOffer = true }
                return
            }
            // Polite side: back out of our own in-flight offer, then accept
            // theirs. Rollback is a LOCAL description operation (W3C
            // `RTCSdpType.rollback`); it is a no-op if our local offer has
            // not landed yet (mid-`createOffer` glare race).
            if session.signalingState == .haveLocalOffer {
                try await session.setLocalDescription(SessionDescription(type: .rollback, sdp: ""))
            }
        }

        try await session.setRemoteDescription(sdp)
        if session.signalingState == .haveRemoteOffer {
            let answer = try await session.createAnswer()
            try await session.setLocalDescription(answer)
            let answerPayload = OfferPayload(sdp: answer.sdp, label: payload.label)
            onLocalAnswer?(answerPayload)
            try signaling.sendAnswer(answerPayload)
        }
        await flushPendingIceCandidates()
    }

    func receiveRemoteAnswer(_ payload: OfferPayload) async throws {
        guard !withLock({ closed }) else { return }
        let sdp = SessionDescription(type: .answer, sdp: payload.sdp)
        try await session.setRemoteDescription(sdp)
        withLock { ignoreRemoteOffer = false }
        await flushPendingIceCandidates()
    }

    // MARK: Data channels

    /// Ensures the bus exists (idempotent). The bus is created eagerly in
    /// `init`, so this is a safety no-op in practice.
    private func ensureDataChannelBus() {
        withLock {
            if _dataChannelBus == nil {
                _dataChannelBus = DataChannelBus(session: session, name: PeerConnectionManager.defaultDataChannelName)
            }
        }
    }

    /// A remote data channel arrived (`ondatachannel`, answerer side): adopt
    /// it into the bus. The remote channel becomes active (only the offerer's
    /// channel is actually negotiated on the wire).
    private func handleRemoteDataChannel(_ channel: DataChannelSession) {
        withLock { _dataChannelBus }?.adoptRemote(channel)
    }

    // MARK: ICE + status plumbing

    private func handleLocalIceCandidate(_ candidate: IceCandidate) {
        let payload = IcePayload(
            candidate: candidate.sdp,
            sdpMid: candidate.sdpMid,
            sdpMLineIndex: Int(candidate.sdpMLineIndex)
        )
        onLocalIceCandidate?(payload)
        try? signaling.sendIce(payload)
    }

    /// Applies one remote candidate on the session (fire-and-forget for the
    /// immediate path; the returned task is awaited by the queue flush so the
    /// state machine is deterministic, mirroring the TS core).
    @discardableResult
    private func addCandidate(_ candidate: IceCandidate) -> Task<Void, Never> {
        Task { [weak self] in
            guard let self else { return }
            do {
                try await self.session.addIceCandidate(candidate)
            } catch {
                // The remote description changed under us (e.g. renegotiation)
                // or the candidate raced an ignored offer: re-queue it; it is
                // retried on the next flush.
                withLock {
                    guard !self.ignoreRemoteOffer else { return }
                    self.pendingIceCandidates.append(candidate)
                }
            }
        }
    }

    /// Applies all queued trickle candidates. Called after a remote
    /// description lands; awaits each add so callers observe a settled state.
    private func flushPendingIceCandidates() async {
        guard session.remoteDescriptionIsSet else { return }
        let queued = withLock { () -> [IceCandidate] in
            let queued = pendingIceCandidates
            pendingIceCandidates.removeAll()
            return queued
        }
        for candidate in queued {
            await addCandidate(candidate).value
        }
    }

    private func negotiationNeeded() {
        Task { [weak self] in
            try? await self?.negotiate()
        }
    }

    private func handleConnectionState(_ state: PeerConnectionState) {
        switch state {
        case .connected:
            status = .connected
        case .failed:
            status = .failed("ICE connection failed")
            if configuration.autoRestartIce {
                Task { [weak self] in
                    try? await self?.restartIce()
                }
            }
        case .disconnected, .closed:
            status = .disconnected
        default:
            status = .connecting
        }
    }

    private func accepts(senderId: String) -> Bool {
        guard let expected = remotePeerId else { return true }
        return senderId == expected
    }
}

File: packages/core/src/peer-connection-manager.ts
/**
 * PeerConnectionManager — mesh core (webrtc-js.md §1.3, §4.2, docs/architecture.md D3).
 *
 * Implements the W3C **perfect negotiation** pattern (polite/impolite), trickle
 * ICE, renegotiation, and ICE restart, plus the engine-side ordering rules:
 *
 *  - **Glare**: concurrent offers are resolved by a deterministic tie-break —
 *    `polite = selfId < remoteId`. The polite peer rolls back its own in-flight
 *    offer and accepts the remote offer; the impolite peer ignores a remote
 *    offer that collides with its own.
 *  - **Trickle ICE (RFC 8838)**: `icecandidate` events are signaled as they
 *    fire; remote candidates are buffered until the matching remote description
 *    has been applied (spec requires `addIceCandidate` after
 *    `setRemoteDescription`).
 *  - **Idempotency**: retransmitted offers/answers are ignored via the SDP
 *    `o=` session-id/session-version (`SdpIdempotencyGuard`).
 *
 * The manager is transport-agnostic: it emits `PeerSignal`s that the caller
 * (e.g. `Room`) forwards over a `SignalingTransport`, and consumes
 * `PeerSignal`s coming from the wire via `handleSignal`.
 */
import type { IcePayload, OfferPayload } from '@mbsks/protocol';
import { SdpIdempotencyGuard } from './sdp.ts';

export type PeerSignal =
  | { type: 'offer'; payload: OfferPayload }
  | { type: 'answer'; payload: OfferPayload }
  | { type: 'ice'; payload: IcePayload };

export interface PeerConnectionManagerOptions {
  /** The platform RTCPeerConnection (browser, werift, wrtc, or fake). */
  pc: RTCPeerConnection;
  /** True if this peer is the polite one (typically `selfId < remoteId`). */
  polite: boolean;
  /** Called with signals that must be forwarded to the remote peer. */
  onSignal?: (signal: PeerSignal) => void;
  onConnectionState?: (state: RTCPeerConnectionState) => void;
  onIceConnectionState?: (state: RTCIceConnectionState) => void;
  onIceGatheringState?: (state: RTCIceGatheringState) => void;
  onDataChannel?: (channel: RTCDataChannel) => void;
  onTrack?: (event: RTCTrackEvent) => void;
  onError?: (error: Error) => void;
  /** Automatically restart ICE when `iceConnectionState` becomes 'failed'. */
  autoRestartIce?: boolean;
  /** Diagnostic logger (no-op by default). */
  debug?: (message: string, data?: unknown) => void;
}

/**
 * A single remote ICE candidate, possibly not yet applied.
 */
interface PendingCandidate {
  candidate: IcePayload;
  attempts: number;
}

export class PeerConnectionManager {
  readonly pc: RTCPeerConnection;
  readonly polite: boolean;

  private readonly onSignal?: (signal: PeerSignal) => void;
  private readonly onConnectionState?: (state: RTCPeerConnectionState) => void;
  private readonly onIceConnectionState?: (state: RTCIceConnectionState) => void;
  private readonly onIceGatheringState?: (state: RTCIceGatheringState) => void;
  private readonly onDataChannel?: (channel: RTCDataChannel) => void;
  private readonly onTrack?: (event: RTCTrackEvent) => void;
  private readonly onError?: (error: Error) => void;
  private readonly autoRestartIce: boolean;
  private readonly debug: (message: string, data?: unknown) => void;

  private readonly sdpGuard = new SdpIdempotencyGuard();
  private readonly pendingCandidates: PendingCandidate[] = [];
  private makingOffer = false;
  private ignoreOffer = false;
  private restartingIce = false;
  private pendingRestart = false;
  private disposed = false;
  private handlersAttached = false;

  constructor(options: PeerConnectionManagerOptions) {
    this.pc = options.pc;
    this.polite = options.polite;
    this.onSignal = options.onSignal;
    this.onConnectionState = options.onConnectionState;
    this.onIceConnectionState = options.onIceConnectionState;
    this.onIceGatheringState = options.onIceGatheringState;
    this.onDataChannel = options.onDataChannel;
    this.onTrack = options.onTrack;
    this.onError = options.onError;
    this.autoRestartIce = options.autoRestartIce ?? true;
    this.debug = options.debug ?? (() => {});
    this.attach();
  }

  // ------------------------------------------------------------------ state

  get signalingState(): RTCSignalingState {
    return this.pc.signalingState;
  }

  get connectionState(): RTCPeerConnectionState {
    return this.pc.connectionState;
  }

  get iceConnectionState(): RTCIceConnectionState {
    return this.pc.iceConnectionState;
  }

  /**
   * All senders on this connection (exposed for the adaptive-quality
   * controller, which reads/updates encodings via `setParameters`).
   */
  getSenders(): RTCRtpSender[] {
    return this.pc.getSenders();
  }

  // ------------------------------------------------------------ event wiring

  private attach(): void {
    if (this.handlersAttached) return;
    this.handlersAttached = true;
    this.pc.onnegotiationneeded = () => {
      this.negotiate().catch((err: unknown) => this.reportError(err, 'negotiationneeded'));
    };
    this.pc.onicecandidate = (event: RTCPeerConnectionIceEvent) => {
      const { candidate } = event;
      if (!candidate) return; // end-of-candidates marker — nothing to signal
      const payload: IcePayload = {
        candidate: candidate.candidate,
        sdpMid: candidate.sdpMid,
        sdpMLineIndex: candidate.sdpMLineIndex ?? null,
      };
      this.debug('ice:local', payload);
      this.onSignal?.({ type: 'ice', payload });
    };
    this.pc.onconnectionstatechange = () => {
      this.debug('connectionState', this.pc.connectionState);
      this.onConnectionState?.(this.pc.connectionState);
    };
    this.pc.oniceconnectionstatechange = () => {
      this.debug('iceConnectionState', this.pc.iceConnectionState);
      this.onIceConnectionState?.(this.pc.iceConnectionState);
      if (this.autoRestartIce && this.pc.iceConnectionState === 'failed') {
        this.restartIce().catch((err: unknown) => this.reportError(err, 'auto-ice-restart'));
      }
    };
    this.pc.onicegatheringstatechange = () => {
      this.onIceGatheringState?.(this.pc.iceGatheringState);
    };
    this.pc.ondatachannel = (event: RTCDataChannelEvent) => {
      this.debug('datachannel:remote', event.channel.label);
      this.onDataChannel?.(event.channel);
    };
    this.pc.ontrack = (event: RTCTrackEvent) => {
      this.debug('track:remote', event.track.kind);
      this.onTrack?.(event);
    };
  }

  private detach(): void {
    if (!this.handlersAttached) return;
    this.handlersAttached = false;
    this.pc.onnegotiationneeded = null;
    this.pc.onicecandidate = null;
    this.pc.onconnectionstatechange = null;
    this.pc.oniceconnectionstatechange = null;
    this.pc.onicegatheringstatechange = null;
    this.pc.ondatachannel = null;
    this.pc.ontrack = null;
  }

  // ------------------------------------------------------------- negotiation

  /**
   * Create an offer and signal it (called from `negotiationneeded` or by the
   * caller after adding/removing tracks). Guards against overlapping offers.
   */
  async negotiate(reason = 'negotiationneeded'): Promise<void> {
    if (this.disposed) return;
    if (this.makingOffer) return;
    this.makingOffer = true;
    this.pendingRestart = false;
    try {
      this.debug('negotiate:start', reason);
      const offer = await this.pc.createOffer();
      await this.pc.setLocalDescription(offer);
      const description = this.pc.localDescription;
      if (!description) {
        this.debug('negotiate:no-local-description');
        return;
      }
      this.debug('negotiate:signal-offer', description.sdp?.slice(0, 40));
      this.onSignal?.({
        type: 'offer',
        payload: { sdp: description.sdp ?? '', label: reason },
      });
    } finally {
      this.makingOffer = false;
    }
  }
  // ------------------------------------------------------------ ICE restart

  /**
   * Restart ICE: `pc.restartIce()` triggers `negotiationneeded`, which
   * produces a new offer with fresh ufrag/pwd (RFC 8445 §9). Falls back to a
   * manual renegotiation on stacks without `restartIce` (e.g. some fakes).
   */
  async restartIce(): Promise<void> {
    if (this.disposed) return;
    if (this.restartingIce) return;
    this.restartingIce = true;
    this.pendingRestart = true;
    try {
      if (typeof this.pc.restartIce === 'function') {
        this.debug('ice:restart');
        this.pc.restartIce();
      } else {
        this.debug('ice:restart-fallback');
      }
      // Give the stack one macrotask to fire negotiationneeded; if it didn't
      // (e.g. some fakes/werift), negotiate manually.
      await new Promise<void>((resolve) => setTimeout(resolve, 0));
      if (this.pendingRestart) {
        await this.negotiate('ice-restart');
      }
    } finally {
      this.restartingIce = false;
    }
  }

  // ---------------------------------------------------------- remote signals

  /** Entry point for any signal arriving from the wire. */
  async handleSignal(signal: PeerSignal): Promise<void> {
    if (this.disposed) return;
    switch (signal.type) {
      case 'offer':
      case 'answer':
        await this.handleRemoteDescription(signal.type, signal.payload.sdp);
        break;
      case 'ice':
        await this.handleRemoteCandidate(signal.payload);
        break;
    }
  }

  /**
   * Apply a remote offer/answer (perfect negotiation + SDP idempotency).
   * Remote ICE candidates are flushed once the description is applied.
   */
  async handleRemoteDescription(type: 'offer' | 'answer', sdp: string): Promise<void> {
    if (this.disposed) return;

    // Idempotency: ignore retransmitted descriptions (same session-id, no newer version).
    if (this.sdpGuard.isDuplicate(type, sdp)) {
      this.debug('sdp:duplicate-ignored', { type, sdp: sdp.slice(0, 40) });
      return;
    }

    if (type === 'offer') {
      // Glare: a remote offer that collides with our in-flight offer.
      const offerCollision = this.makingOffer || this.pc.signalingState !== 'stable';
      this.ignoreOffer = !this.polite && offerCollision;
      if (this.ignoreOffer) {
        this.debug('glare:ignored-remote-offer', { polite: this.polite });
        return;
      }
      if (this.polite && offerCollision) {
        // Back out of our own in-flight offer and accept theirs.
        await this.rollback();
      }
    }

    await this.pc.setRemoteDescription({ type, sdp });
    this.sdpGuard.record(type, sdp);

    if (type === 'offer') {
      await this.pc.setLocalDescription();
      const description = this.pc.localDescription;
      if (description) {
        this.onSignal?.({ type: 'answer', payload: { sdp: description.sdp ?? '' } });
      }
    } else {
      this.ignoreOffer = false;
    }

    await this.flushPendingCandidates();
  }

  /**
   * Handle a trickle ICE candidate. Candidates are buffered until the remote
   * description is applied (spec requires `addIceCandidate` after
   * `setRemoteDescription`), then flushed.
   */
  async handleRemoteCandidate(candidate: IcePayload): Promise<void> {
    if (this.disposed) return;
    if (candidate.candidate === '') {
      // End-of-candidates marker: nothing to add.
      return;
    }
    if (!this.pc.remoteDescription) {
      this.pendingCandidates.push({ candidate, attempts: 0 });
      return;
    }
    await this.addCandidate(candidate);
  }

  private async addCandidate(candidate: IcePayload): Promise<void> {
    try {
      await this.pc.addIceCandidate({
        candidate: candidate.candidate,
        sdpMid: candidate.sdpMid ?? undefined,
        sdpMLineIndex: candidate.sdpMLineIndex ?? undefined,
      });
    } catch (err) {
      if (this.ignoreOffer) {
        this.debug('ice:ignored-after-ignored-offer', candidate);
        return;
      }
      // Stacks throw InvalidStateError if the remote description changed under
      // us; re-queue and retry on the next flush.
      this.pendingCandidates.push({ candidate, attempts: 1 });
      this.debug('ice:add-failed-queued', { candidate, error: err });
    }
  }

  private async flushPendingCandidates(): Promise<void> {
    if (!this.pc.remoteDescription) return;
    const queue = [...this.pendingCandidates];
    this.pendingCandidates.length = 0;
    for (const entry of queue) {
      if (entry.attempts > 0) {
        // A previously failed candidate: retry once; if it fails again, drop it.
        try {
          await this.pc.addIceCandidate({
            candidate: entry.candidate.candidate,
            sdpMid: entry.candidate.sdpMid ?? undefined,
            sdpMLineIndex: entry.candidate.sdpMLineIndex ?? undefined,
          });
        } catch (err) {
          this.debug('ice:retry-failed-dropped', { candidate: entry.candidate, error: err });
        }
        continue;
      }
      await this.addCandidate(entry.candidate);
    }
  }

  /** Roll back an in-flight local offer (polite peer during glare). */
  private async rollback(): Promise<void> {
    if (this.pc.signalingState !== 'have-local-offer') return;
    try {
      await this.pc.setLocalDescription({ type: 'rollback' } as RTCSessionDescriptionInit);
    } catch (err) {
      // Some stacks (werift/wrtc) lack rollback; the remote offer will fail to
      // apply and surface via onError instead of being silently mishandled.
      this.debug('rollback:unsupported', err);
      throw err;
    }
  }

  // --------------------------------------------------------------- lifecycle

  /** Detach handlers and close the underlying peer connection. */
  close(): void {
    if (this.disposed) return;
    this.disposed = true;
    this.detach();
    try {
      this.pc.close();
    } catch (err) {
      this.debug('close:error', err);
    }
    this.pendingCandidates.length = 0;
  }

  /** Detach handlers without closing the pc (owner manages lifecycle). */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    this.detach();
    this.pendingCandidates.length = 0;
  }

  private reportError(err: unknown, context: string): void {
    const error = err instanceof Error ? err : new Error(String(err));
    this.debug(`error:${context}`, error);
    this.onError?.(error);
  }
}
